Drivers: comparator: silabs_acmp implementation

silabs-chgalant opened the pull request:
This pull request does the following:

- Enables the hal_silabs ACMP module to be compiled for series 2.
- Adds a generated bindings header to aid with acmp input selection
- Defines a compatible binding which can be used for series 2 parts that have an acmp peripheral
- Instantiates the xg2x parts with an acmp peripheral with an acmp node in their respective part dts
- Implements the comparator driver compatible with silabs,acmp
- Adds support for the comparator driver in the relevant series 2 development kits and radio boards.
- Extends the comparator gpio_loopback tests for xg24_dk2601b and xg29_rb4412a boards.
- Extends comparator build testing for all current series 2 boards that support the comparator.

The `input-positive` and `input-negative` properties of the silabs,acmp compatible binding are part specific, therefore binding headers are generated per part (as per @asmellby's feedback). However, since most of the input values are common between parts, and there is no overlapping of input values, it was decided (as per @jerome-pouiller's feedback) that we compile the total set of possible ACMP inputs into a single bindings header for the silabs_acmp. It would therefore be up to the user to reference their part's design book when selecting the `input-positive` and `input-negative` values for their application.

silabs-chgalant committed:
drivers: comparator: implement the silabs_acmp compatible driver

This implements the comparator driver for silabs acmp peripherals
using the silabs,acmp compatible binding.

Signed-off-by: Christian Galante <christian.galante@silabs.com>

silabs-chgalant committed:
dts: bindings: comparator: define silabs,acmp bindings

Defines bindings that are compatible with silabs acmp.
`input-positive` and `input-negative` are required
properties to be configured by an application.
It is recommended to use the bindings generated in
`include/zephyr/dt-bindings/comparator/silabs_acmp.h`
and reference your part's design book when configuring
values for these properties.

Signed-off-by: Christian Galante <christian.galante@silabs.com>

silabs-chgalant committed:
dts: arm: silabs: instantiate acmp nodes for xg2x parts

Defines an acmp node for xg21, xg23, xg24, xg27 and xg29
parts, which are all compatible with the silabs,acmp
binding.

Signed-off-by: Christian Galante <christian.galante@silabs.com>

silabs-chgalant committed:
soc: silabs: add configuration to enable acmp module

The ACMP module is enabled with this change in the gecko
and simplicity_sdk hal libraries for development on silabs
devices. This is a pre-requisite for implementing the
comparator driver that is compatible with silabs acmp
peripherals.

Signed-off-by: Christian Galante <christian.galante@silabs.com>

silabs-chgalant committed:
module: hal_silabs: enable acmp module in series 2 hal

This enables silabs series 2 hal source files for the silabs
acmp module to be compiled when the SOC_SILABS_ACMP
configuration is selected.

Signed-off-by: Christian Galante <christian.galante@silabs.com>

silabs-chgalant committed:
boards: silabs: update board support for comparator driver

The xg24_dk2601b, xg24_ek2703a, xg27_dk2602a, xg23_rb4210a,
xg24_rb4187c and xg29_rb4412a board yaml files were updated
to support the comparator driver. Documentation hardware
support was updated for each of these boards as well to
mention support for the ACMP hardware block and comparator
driver.

Signed-off-by: Christian Galante <christian.galante@silabs.com>

asmellby commented on dts/bindings/comparator/silabs,acmp.yaml:
```diff
@@ -0,0 +1,101 @@
+      - "VSENSE11DIV4LP"
+    default: "VSENSE01DIV4"
+
+  neg-select:
```
ADC binding uses `input-negative`, should we consider that name for ACMP too?

asmellby commented on dts/bindings/comparator/silabs,acmp.yaml:
```diff
@@ -0,0 +1,101 @@
+      - "REDUCED"
+    default: "FULL"
+
+  pos-select:
```
ADC binding uses `input-positive`, should we consider that name for ACMP too?

asmellby commented on drivers/comparator/comparator_silabs_acmp.c:
```diff
@@ -0,0 +1,184 @@
+		.irq_init = ACMP_IRQ_HANDLER_INIT_FN_SYM(inst),                                    \
+		.init = ACMP_DT_INST_CONFIG_INIT(inst),                                            \
+		.negsel =                                                                          \
+			_CONCAT(_ACMP_INPUTCTRL_NEGSEL_, DT_INST_STRING_TOKEN(inst, neg_select)),  \
```
If the binding is changed to an integer, it can be directly inserted here using `DT_INST_PROP()`.

jerome-pouiller commented on drivers/comparator/CMakeLists.txt:
In your commit log, `.com` is missing at the end of `<christian.galante@silabs>`.

silabs-chgalant replied:
Woops! Thanks for catching that.

jerome-pouiller commented on boards/silabs/dev_kits/xg24_dk2601b/xg24_dk2601b.yaml:
```diff
@@ -15,6 +15,7 @@ supported:
   - dma
   - watchdog
   - clock_control
+  - comparator
```
Probably we want to propagate this change to `xg24_rb4187c` and `xg24_ek2703a`?

BTW, I think we can also declare it in xg21, xg23 and xg27.

silabs-chgalant replied:
Yes indeed. I will be generating device specific bindings for each series 2 device that has an acmp, and at that point it will be possible to propagate the comparator support to each board. I can also instantiate an acmp node in each of the device tree sources for xg21, xg23 and xg27.

silabs-chgalant commented on drivers/comparator/comparator_silabs_acmp.c:
```diff
@@ -0,0 +1,184 @@
+	void *user_data;
+};
+
+#define ACMP_DT_INST_CONFIG_INIT(inst)                                                             \
```
I think inlining makes sense. Thanks!

silabs-chgalant commented on modules/hal_silabs/gecko/CMakeLists.txt:
```diff
@@ -82,6 +82,7 @@ zephyr_compile_definitions(
 )
 
 zephyr_library_sources(                                    ${EMLIB_DIR}/src/em_system.c)
+zephyr_library_sources_ifdef(CONFIG_SOC_SILABS_ACMP        ${EMLIB_DIR}/src/em_acmp.c)
```
I will remove this line, the support will be for series 2 devices.

silabs-chgalant commented on drivers/comparator/comparator_silabs_acmp.c:
```diff
@@ -0,0 +1,184 @@
+	switch (trigger) {
+	case COMPARATOR_TRIGGER_BOTH_EDGES:
+		ACMP_IntEnable(config->reg, ACMP_IEN_RISE | ACMP_IEN_FALL);
+		data->interrupt_mask = ACMP_IEN_RISE | ACMP_IEN_FALL;
```
Yes indeed, thanks for catching that!

silabs-chgalant committed:
include: zephyr: dt-bindings: comparator: silabs: generate acmp

Generates bindings that define acmp inputs. These bindings can be
used in devicetree sources to configure the acmp's
input-positive and input-negative properties. Depending on the
part, some inputs may not be available as positive and/or
negative inputs to the ACMP, therefore always reference your
part's design book when selecting these inputs.

Signed-off-by: Christian Galante <christian.galante@silabs.com>

silabs-chgalant reviewed (commented):
This commit message should not be specific to xg24, will need to update this.

silabs-chgalant commented on boards/silabs/dev_kits/xg24_dk2601b/xg24_dk2601b.yaml:
Will need to update the commit message since these changes are no longer just for xg24.

silabs-chgalant commented on dts/bindings/comparator/silabs,acmp.yaml:
```diff
@@ -0,0 +1,85 @@
+           input-negative = <ACMP_INPUT_NEGATIVE_VREFDIVAVDD>;
+    };
+
+  Note that the ACMP_INPUT_POSITIVE and ACMP_INPUT_NEGATIVE bindings are device
```
Meant to say "the input-positive and input-negative properties" instead of "the ACMP_INPUT_POSITIVE and ACMP_INPUT_NEGATIVE bindings".

asmellby commented on include/zephyr/dt-bindings/acmp/silabs/xg21-acmp.h:
Directory should match driver/binding directory name: `dt-bindings/comparator`, not `dt-bindings/acmp`

asmellby commented on include/zephyr/dt-bindings/acmp/silabs/xg21-acmp.h:
```diff
@@ -0,0 +1,167 @@
+#ifndef ZEPHYR_INCLUDE_DT_BINDINGS_SILABS_XG21_ACMP_H_
+#define ZEPHYR_INCLUDE_DT_BINDINGS_SILABS_XG21_ACMP_H_
+
+#include <zephyr/dt-bindings/dt-util.h>
```
`dt-util.h` doesn't seem to be used in this file, include can be removed.

asmellby commented on dts/arm/silabs/efr32bg27.dtsi:
```diff
@@ -35,6 +35,14 @@
 		};
 	};
+
+	acmp0: acmp@5a008000 {
```
ACMP node should be inside `soc` node.

asmellby commented on boards/silabs/dev_kits/xg24_dk2601b/xg24_dk2601b.yaml:
Commenting on this file since I can't comment on the commit itself: Please also add comparator to the table of supported drivers in the board documentation at `doc/index.rst` for each board.

jerome-pouiller commented on include/zephyr/dt-bindings/acmp/silabs/xg21-acmp.h:
```diff
@@ -0,0 +1,167 @@
+
+#include <zephyr/dt-bindings/dt-util.h>
+
+#define ACMP_INPUT_POSITIVE_VSS               0
```
Does it make sense to one set of values for Positive and another for Negative? Ditto for the chipset, does it make sense to specific set of value of each chipset while they are mostly identical?

I believe this list would fit all:
```
#define ACMP_INPUT_VSS               0
#define ACMP_INPUT_VREFDIVAVDD       16
#define ACMP_INPUT_VREFDIVAVDDLP     17
#define ACMP_INPUT_VREFDIV1V25       18
#define ACMP_INPUT_VREFDIV1V25LP     19
#define ACMP_INPUT_VREFDIV2V5        20
#define ACMP_INPUT_VREFDIV2V5LP      21
#define ACMP_INPUT_VSENSE01DIV4      32
#define ACMP_INPUT_VSENSE01DIV4LP    33
#define ACMP_INPUT_VSENSE11DIV4      34
#define ACMP_INPUT_VSENSE11DIV4LP    35
#define ACMP_INPUT_CAPSENSE          48
#define ACMP_INPUT_VDAC0OUT0         64
#define ACMP_INPUT_VDAC0OUT1         65
#define ACMP_INPUT_VDAC1OUT0         66
#define ACMP_INPUT_VDAC1OUT1         67
#define ACMP_INPUT_EXTPA             80
#define ACMP_INPUT_EXTPB             81
#define ACMP_INPUT_EXTPC             82
#define ACMP_INPUT_EXTPD             83
#define ACMP_INPUT_PA0               128
#define ACMP_INPUT_PA1               129
#define ACMP_INPUT_PA2               130
#define ACMP_INPUT_PA3               131
#define ACMP_INPUT_PA4               132
#define ACMP_INPUT_PA5               133
#define ACMP_INPUT_PA6               134
#define ACMP_INPUT_PA7               135
#define ACMP_INPUT_PA8               136
#define ACMP_INPUT_PA9               137
#define ACMP_INPUT_PA10              138
#define ACMP_INPUT_PA11              139
#define ACMP_INPUT_PA12              140
#define ACMP_INPUT_PA13              141
#define ACMP_INPUT_PA14              142
#define ACMP_INPUT_PA15              143
#define ACMP_INPUT_PB0               144
#define ACMP_INPUT_PB1               145
#define ACMP_INPUT_PB2               146
#define ACMP_INPUT_PB3               147
#define ACMP_INPUT_PB4               148
#define ACMP_INPUT_PB5               149
#define ACMP_INPUT_PB6               150
#define ACMP_INPUT_PB7               151
#define ACMP_INPUT_PB8               152
#define ACMP_INPUT_PB9               153
#define ACMP_INPUT_PB10              154
#define ACMP_INPUT_PB11              155
#define ACMP_INPUT_PB12              156
#define ACMP_INPUT_PB13              157
#define ACMP_INPUT_PB14              158
#define ACMP_INPUT_PB15              159
#define ACMP_INPUT_PC0               160
#define ACMP_INPUT_PC1               161
#define ACMP_INPUT_PC2               162
#define ACMP_INPUT_PC3               163
#define ACMP_INPUT_PC4               164
#define ACMP_INPUT_PC5               165
#define ACMP_INPUT_PC6               166
#define ACMP_INPUT_PC7               167
#define ACMP_INPUT_PC8               168
#define ACMP_INPUT_PC9               169
#define ACMP_INPUT_PC10              170
#define ACMP_INPUT_PC11              171
#define ACMP_INPUT_PC12              172
#define ACMP_INPUT_PC13              173
#define ACMP_INPUT_PC14              174
#define ACMP_INPUT_PC15              175
#define ACMP_INPUT_PD0               176
#define ACMP_INPUT_PD1               177
#define ACMP_INPUT_PD2               178
#define ACMP_INPUT_PD3               179
#define ACMP_INPUT_PD4               180
#define ACMP_INPUT_PD5               181
#define ACMP_INPUT_PD6               182
#define ACMP_INPUT_PD7               183
#define ACMP_INPUT_PD8               184
#define ACMP_INPUT_PD9               185
#define ACMP_INPUT_PD10              186
#define ACMP_INPUT_PD11              187
#define ACMP_INPUT_PD12              188
#define ACMP_INPUT_PD13              189
#define ACMP_INPUT_PD14              190
#define ACMP_INPUT_PD15              191
```
(FYI, I generated this list with: `< 37f895cfd4bc003e70ae502acb58c3047ba45ee9.patch sed -n -e 's/^+//' -e 's/_POSITIVE//p' -e 's/_NEGATIVE//p' | sort -n -k 3 | uniq`. I have only renamed `ACMP_INPUT_VDACOUT0` and `ACMP_INPUT_VDACOUT1`)

silabs-chgalant replied:
There's also the fact that CAPSENSE is only permitted as a NEGATIVE input, and some inputs are not available on some chipsets (i.e. VDAC is not available on all parts this was generated for). 

I think from a user's perspective it will be a lot easier to look at the list of available positive inputs and negative inputs for the part you're working with and select those than it is to select an input and then debug an invalid configuration afterwards.

jerome-pouiller replied:
In my humble (and subjective) opinion, we should Keep Things Simple and Stupid.

silabs-chgalant replied:
I don't have a hard stance on this, we can compile all of the possible inputs for boards into a singular bindings header and in that case it will be up to the user to read the design book for the part they are using to know which inputs are not valid.

jerome-pouiller commented on include/zephyr/dt-bindings/acmp/silabs/xg21-acmp.h:
```diff
@@ -0,0 +1,167 @@
+
+#include <zephyr/dt-bindings/dt-util.h>
+
+#define ACMP_INPUT_POSITIVE_VSS               0
```
Nitpick: when constants are > 30, I tend to prefer hexadecimal.

jerome-pouiller commented on drivers/comparator/comparator_silabs_acmp.c:
```diff
@@ -0,0 +1,184 @@
+		.init.enable = true,                                                               \
+		.input_negative = DT_INST_PROP(inst, input_negative),                              \
+		.input_positive = DT_INST_PROP(inst, input_positive),                              \
+	};                                                                                         \
```
It's much better now :)

silabs-chgalant commented on include/zephyr/dt-bindings/acmp/silabs/xg21-acmp.h:
```diff
@@ -0,0 +1,167 @@
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * This file was generated by the script gen_acmp.py in the hal_silabs module.
```
Yes, I will open a PR in `hal_silabs` for that script shortly, then update the reference in west.yml once that gets merged.

silabs-chgalant replied:
I've opened the PR for `hal_silabs` here https://github.com/zephyrproject-rtos/hal_silabs/pull/80 (cc @asmellby)

jerome-pouiller commented:
I believe this PR is no more a draft.

bjarki-andreasen commented:
Looks good :) could you extend the test suite https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/drivers/comparator/gpio_loopback with the new comparator and test it out locally as well? and extend https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/drivers/build_all/comparator as well? That allows CI to build various configurations for various boards too :)

silabs-chgalant commented:
> Looks good :) could you extend the test suite https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/drivers/comparator/gpio_loopback with the new comparator and test it out locally as well? and extend https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/drivers/build_all/comparator? That allows CI to build various configurations for various boards too :)

Thank you for the review!

I'm in the process of extending these tests, I will let you know when this work has been pushed.

silabs-chgalant committed:
tests: drivers: comparator: gpio_loopback: extend to xg24 and xg29 boards

Add board overlays to execute the comparator gpio_loopback testsuite on
xg24_dk2601b and xg29_rb4412a. The testsuite was executed and was seen
to pass on both boards locally with twister.

Signed-off-by: Christian Galante <christian.galante@silabs.com>

silabs-chgalant committed:
tests: drivers: build_all: comparator: add silabs_acmp test configurations

Add build test configurations for the silabs_acmp comparator. Enable this
build testing on all series 2 silabs boards that currently support the
comparator driver.

Signed-off-by: Christian Galante <christian.galante@silabs.com>

jerome-pouiller reviewed (approved):
Nice !

asmellby commented on drivers/comparator/comparator_silabs_acmp.c:
```diff
@@ -0,0 +1,288 @@
+			}                                                                          \
+			break;                                                                     \
+		default:                                                                           \
+			return -EIO;                                                               \
```
Should this be `return 0` or `break`? If it doesn't match any port, it means it doesn't need ABUS allocation, it's not an error AFAICT.

silabs-chgalant replied:
Yes indeed this is what should have been returned. Good catch!

asmellby commented on drivers/comparator/comparator_silabs_acmp.c:
```diff
@@ -0,0 +1,288 @@
+	}
+
+	/* Allocate ABUS for GPIO inputs */
+	config->input_abus_allocate(config->input_positive);
```
This function returns an error code that should be checked. Failure to allocate an ABUS can happen if 3 peripherals try to use the same ABUS (e.g. ACMP0, ACMP1 and IADC0 all on port A even pins).

silabs-chgalant replied:
Due to the abus allocation being done through pinctrl, I'll resolve this comment. Thanks!

bjarki-andreasen commented:
I will wait to approve until comments by @asmellby are addressed :) They look blocking though they are "just comments" :)

jerome-pouiller commented on drivers/comparator/comparator_silabs_acmp.c:
```diff
@@ -0,0 +1,288 @@
+			if (ACMP_INPUT_PIN(input) % 2 == 0) {                                      \
+				if ((GPIO->ABUSALLOC & _GPIO_ABUSALLOC_AEVEN0_MASK) ==             \
+				    GPIO_ABUSALLOC_AEVEN0_TRISTATE) {                              \
+					GPIO->ABUSALLOC |= GPIO_ABUSALLOC_AEVEN0_ACMP##inst;       \
```
Sorry, I missed this change during my review (Github is not great to follow versions of a PR).

This function directly access to the GPIO device. I believe this should goes in the  pinctl driver.

`inst` is just the index of declaration in the device tree. You can't rely on it.

asmellby replied:
Pinctrl is for digital pin control, analog pins are completely different. The purpose of this is to allocate one of 6 ABUSes to the peripheral, the peripheral then directly configures its own pins using possel/negsel registers separately. This function should really exist in the HAL, but it doesn't (yet). I suggested to temporarily put it in the driver until the HAL provides the functionality.

Agree that `inst` can't really be used here, the register value must instead be created based on the actual peripheral instance, either by comparing `DT_INST_REG` to the peripheral base pointer or some other solution. Anyway, the function can be promoted to a "real" function rather than a macro-templated one, since it should no longer rely on instance-specific data.

jerome-pouiller replied:
I am still not very comfortable with this approach. Here `GPIO` is another hardware block. We try to have 1 driver == 1 hardware block == 1 Zephyr API. This rule is not always easy to apply. However this case seems obvious enough.

It seems it is rather usual the ADCs rely on pinctrl for the configuration of the analog pins (I have browsed a bit the results of `git grep pinctrl.*adc`). 

I agree if we add `pinctrl` in the loop, the user will probably have to repeat himself in pinctrl and in acmp. However, I believe this is more transparent.

jerome-pouiller replied:
Yeah, I expected something like:
```
diff --git i/include/zephyr/dt-bindings/pinctrl/silabs-pinctrl-dbus.h w/include/zephyr/dt-bindings/pinctrl/silabs-pinctrl-dbus.h
index 9ad02a5cd0f..1a659e81eb2 100644
--- i/include/zephyr/dt-bindings/pinctrl/silabs-pinctrl-dbus.h
+++ w/include/zephyr/dt-bindings/pinctrl/silabs-pinctrl-dbus.h
@@ -28,6 +28,7 @@
 #define SILABS_PINCTRL_HAVE_EN_MASK     0x00040000UL
 #define SILABS_PINCTRL_EN_BIT_MASK      0x00F80000UL
 #define SILABS_PINCTRL_ROUTE_MASK       0x1F000000UL
+#define SILABS_PINCTRL_IS_ANALOG        0x80000000UL

 #define SILABS_DBUS(port, pin, periph_base, en_present, en_bit, route)                             \
        (FIELD_PREP(SILABS_PINCTRL_GPIO_PORT_MASK, port) |                                         \
@@ -37,4 +38,10 @@
         FIELD_PREP(SILABS_PINCTRL_EN_BIT_MASK, en_bit) |                                          \
         FIELD_PREP(SILABS_PINCTRL_ROUTE_MASK, route))

+#define SILABS_ABUS(port, pin, func)                             \
+        (SILABS_PINCTRL_IS_ANALOG |                              \
+         FIELD_PREP(SILABS_PINCTRL_GPIO_PORT_MASK, port) |       \
+         FIELD_PREP(SILABS_PINCTRL_GPIO_PIN_MASK, pin) |         \
+         FIELD_PREP(SILABS_PINCTRL_PERIPH_BASE_MASK, func) )
+
 #endif /* ZEPHYR_INCLUDE_DT_BINDINGS_PINCTRL_SILABS_PINCTRL_DBUS_H_ */```

asmellby replied:
To frame my position in a different way: The selection being made in GPIO registers is not a pin selection, it's a _bus_ selection. For analog peripherals, the pin selection happens inside the peripheral itself, which is different from digital peripherals, that have centralized pin selection.

I don't disagree that management should be centralized, I just don't think pinctrl is the right interface for analog bus selection (unless you move the entire pin selection there, which means the pinctrl driver needs to paint registers across all the analog peripherals, which is equally "unclean" from a 1 driver == 1 peripheral point of view).

asmellby replied:
`SILABS_PINCTRL_PERIPH_BASE_MASK` would need a completely different definition for analog peripherals, considering that they are not entries within GPIO DBUS.

asmellby replied:
I wrote a proposal for an implementation of ABUS in the pinctrl driver, feedback welcome: https://github.com/zephyrproject-rtos/zephyr/pull/84462

kartben commented:
this will need a rebase due to merge conflicts @silabs-chgalant

jerome-pouiller commented:
@silabs-chgalant, #84462 has been merged. Can you rebase your work?

silabs-chgalant commented:
> @silabs-chgalant, #84462 has been merged. Can you rebase your work?

Thanks! I've rebased locally and I'm now making the necessary changes to integrate the pinctrl analog bus allocation into the comparator driver.

asmellby commented on include/zephyr/dt-bindings/comparator/silabs/silabs-acmp.h:
I would place this file one directory higher, no need for the `silabs` directory when we don't have several files for each device.

silabs-chgalant replied:
Makes sense!

asmellby commented on dts/bindings/comparator/silabs,acmp.yaml:
```diff
@@ -0,0 +1,147 @@
+    &pinctrl {
+            acmp0_default: acmp0_default {
+                    group0 {
+                            silabs,analog-bus = <>;
```
Example should show usage of an actual ABUS macro IMO (and include the pinctrl header)

silabs-chgalant replied:
There are two examples in this documentation, the code block lower down shows the abus allocation. I feel that we need to show both examples since when there is no GPIO input, the pinctrl configuration still needs to configure an empty list for silabs,analog-bus.

asmellby commented on drivers/comparator/comparator_silabs_acmp.c:
```diff
@@ -0,0 +1,203 @@
+	}
+
+	err = pinctrl_apply_state(config->pincfg, PINCTRL_STATE_DEFAULT);
+	if (err < 0) {
```
This needs to consider `-ENOENT` as a success value, since no pinctrl configuration will exist if no pins are measured (only supplies).

silabs-chgalant replied:
Is this maybe why I was running into an issue when I didn't have this empty pinctrl configuration?
`
&pinctrl {
            acmp0_default: acmp0_default {
                    group0 {
                            silabs,analog-bus = <>;
                    };
          };
};
`

silabs-chgalant replied:
I just did a quick test, considering `-ENOENT` as successful indeed makes the "no pin" configuration function without the empty pinctrl configuration. I'll clean up the overlays and documentation regarding that and modify the conditional to `if (err < 0 && err != -ENOENT)`

silabs-chgalant commented on drivers/comparator/comparator_silabs_acmp.c:
```diff
@@ -0,0 +1,203 @@
+/*
+ * Copyright (c) 2025 Silicon Labs
```
I'll push a fix to this for all of the files. Thanks for catching that!

silabs-chgalant commented:
@Martinhoff-maker @jerome-pouiller @bjarki-andreasen, I pushed a fix for the copyright declarations.

jerome-pouiller commented on drivers/comparator/Kconfig.silabs_acmp:
```diff
@@ -0,0 +1,12 @@
+# Copyright (c) 2025 Silicon Labs Inc.
```
In fact, we tend to use "Silicon Laboratories Inc." 🙄

jhedberg commented:
@bjarki-andreasen looks like this should be ready to go in - just missing your review/approval. Thanks!

bjarki-andreasen reviewed (approved):
Nice work!
